rewrite-java-tck/src/main/java/org/openrewrite/java/tree/UnnamedVariableTest.java

```diff
@@ -0,0 +1,325 @@
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.tree;
+
+import org.junit.jupiter.api.Test;
+import org.openrewrite.java.MinimumJava25;
+import org.openrewrite.test.RewriteTest;
+
+import static org.openrewrite.java.Assertions.java;
+
+@MinimumJava25
+class UnnamedVariableTest implements RewriteTest {
+
+    // Based on https://docs.oracle.com/en/java/javase/21/language/unnamed-variables-and-patterns.html
+
+    @Test
+    void unnamedVariableInForLoop() {
+        rewriteRun(
+          java(
+            """
+              import java.util.List;
+
+              class Test {
+                  void test(List<String> list) {
+                      int total = 0;
+                      for (var _ : list) {
+                          total++;
+                      }
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    void unnamedVariableInCatchBlock() {
+        rewriteRun(
+          java(
+            """
+              class Test {
+                  void test() {
+                      try {
+                          int x = Integer.parseInt("123");
+                      } catch (NumberFormatException _) {
+                          System.out.println("Bad number format");
+                      }
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    void unnamedVariableInTryWithResources() {
+        rewriteRun(
+          java(
+            """
+              import java.io.FileReader;
+
+              class Test {
+                  void test(String path) {
+                      try (var _ = new FileReader(path)) {
+                          System.out.println("Critical section");
+                      }
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    void unnamedVariableInLambda() {
+        rewriteRun(
+          java(
+            """
+              import java.util.function.BiFunction;
+
+              class Test {
+                  void test() {
+                      BiFunction<Integer, Integer, Integer> func = (x, _) -> x * 2;
+                  }
+              }
+              """
+          )
+        );
+    }
+
+    @Test
+    void multipleUnnamedVariablesInLambda() {
+        rewriteRun(
+          java(
+            """
+              interface TriConsumer<T, U, V> {
+                  void accept(T t, U u, V v);
+              }
+
+              class Test {
+                  void test() {
+                      TriConsumer<String, String, String> consumer = (_, _, third) ->
+                          System.out.println(third);
```

Comment:
I wonder if we should add some after recipe check here to verify that the first and second `_` are not semantically equal.

Reply:
Not a bad idea

Reply:
I meant a slightly different assertion as now added in https://github.com/openrewrite/rewrite/pull/5971/commits/3081b5daf5a83acb1957a46f60953a769f4c4bb9 ; that's made the test fail, and I think it would be good to at least explore a fix.